File: 60_permutation_sequence.py
def permute_until_k(arr, k, start=0, count=[0], result=[None]):
    # Stop early if we already found the result
    if result[0] is not None:
        return

    if start == len(arr) - 1:
        count[0] += 1
        if count[0] == k:
            result[0] = arr.copy()
        return

    for i in range(start, len(arr)):
        arr[start], arr[i] = arr[i], arr[start]  # Swap
        
        permute_until_k(arr, k, start + 1, count, result)
        
        arr[start], arr[i] = arr[i], arr[start]  # Backtrack
        
        if result[0] is not None:  # Extra stop after recursion returns
            return

# ---------- User Input ----------
n = int(input("Enter n (1 to 9): "))
k = int(input("Enter k (1 to n!): "))

nums = list(range(1, n + 1))  # Create list [1, 2, 3, ..., n]

result = [None]
permute_until_k(nums, k, result=result)

if result[0]:
    print("The", k, "th permutation is:", ''.join(map(str, result[0])))
else:
    print("Invalid k value!")

File: 60_permutation_sequence.js
function permuteUntilK(
  arr,
  k,
  start = 0,
  count = { value: 0 },
  result = { value: null }
) {
  if (result.value !== null) return; // Stop early if found

  if (start === arr.length - 1) {
    count.value += 1;
    if (count.value === k) {
      result.value = arr.slice(); // Copy the array
    }
    return;
  }

  for (let i = start; i < arr.length; i++) {
    // Swap
    [arr[start], arr[i]] = [arr[i], arr[start]];

    permuteUntilK(arr, k, start + 1, count, result);

    // Backtrack
    [arr[start], arr[i]] = [arr[i], arr[start]];

    if (result.value !== null) return; // Extra stop after returning
  }
}

// --------- User Input (Node.js) ---------

const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

readline.question("Enter n (1 to 9): ", (nInput) => {
  readline.question("Enter k (1 to n!): ", (kInput) => {
    const n = parseInt(nInput);
    const k = parseInt(kInput);

    if (n < 1 || n > 9) {
      console.log("n must be between 1 and 9.");
      readline.close();
      return;
    }

    const nums = [];
    for (let i = 1; i <= n; i++) nums.push(i);

    const result = { value: null };
    permuteUntilK(nums, k, 0, { value: 0 }, result);

    if (result.value) {
      console.log(`The ${k}th permutation is: ${result.value.join("")}`);
    } else {
      console.log("Invalid k value!");
    }

    readline.close();
  });
});
